src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2023 Snowflake Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.snowflake.kafka.connector.internal.streaming;
+
+import static com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig.SNOWPIPE_STREAMING_FILE_VERSION;
+import static net.snowflake.ingest.utils.ParameterProvider.BLOB_FORMAT_VERSION;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig;
+import com.snowflake.kafka.connector.Utils;
+import com.snowflake.kafka.connector.internal.KCLogger;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClientFactory;
+import net.snowflake.ingest.utils.SFException;
+import org.apache.kafka.connect.errors.ConnectException;
+
+/** Singleton that provides the streaming client */
+public class StreamingClientProvider {
+  public static final StreamingClientProvider streamingClientProvider =
+      new StreamingClientProvider();
+
+  /**
+   * Checks if the client is valid by doing a null check, ensuring it is open and has a name
+   *
+   * @param client The client to validate
+   * @return If the client is not null, open and has a name
+   */
+  public static boolean isClientValid(SnowflakeStreamingIngestClient client) {
+    return client != null && !client.isClosed() && client.getName() != null;
+  }
+
+  /**
+   * ONLY FOR TESTING - to get a provider with injected properties
+   *
+   * @param createdClientId the number of times a client has been created
+   * @param connectorConfig the connector config
+   * @param client the injected streaming client
+   * @return a provider with the injected properties
+   */
+  @VisibleForTesting
+  public static StreamingClientProvider injectStreamingClientProviderForTests(
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    return new StreamingClientProvider(createdClientId, connectorConfig, client);
+  }
+
+  private static final KCLogger LOGGER = new KCLogger(Utils.class.getName());
+  private static final String STREAMING_CLIENT_PREFIX_NAME = "KC_CLIENT_";
+
+  private int createdClientId;
```

Comment:
Should this be threadsafe? 
Use `AtomicInteger` instead.

Reply:
Good call, using AtomicInteger